Base node block sync RPC service

sdbondi opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
<!--- Describe your changes in detail -->
RPC service for faster block sync. Allows streaming of full blocks,
headers and a method for chain split detection. The client-side is next
up.

- Added BlockchainStateService. This service should eventually be the only service
  which provides a query interface to the blockchain database for other
  services.
- Added BaseNodeSyncRpcService. This RPC service exposes `sync_blocks`,
  `sync_headers`, `fetch_header_by_height` and `find_chain_split`. These
  functions will be used in the improved block sync.
- Added `NotFound` status to RpcStatus
- Added `RpcStatus::log_internal_error()` convenience function to make
  dealing with misc errors in RPC handler functions easier.
- Added `spawn_until_shutdown` method to the service initialiser context 
   that spawns a task until the shutdown is triggered. This removes the need for   
   implementers to worry about how to implement shutdown in initialisers.
- Cleaned up some test helpers
- Added unit tests for the Base node sync RPC handlers
- Added unit tests for the BlockchainStateService
- Remove unused `tokio-test` dependency

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
Described here: https://github.com/tari-project/tari/issues/2197#issuecomment-708502628
Ref #2113

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
Extensive unit tests

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
* [ ] Bug fix (non-breaking change which fixes an issue)
* [x] New feature (non-breaking change which adds functionality)
* [ ] Breaking change (fix or feature that would cause existing functionality to change)
* [ ] Feature refactor (No new feature or functional changes, but performance or technical debt improvements)
* [x] New Tests
* [ ] Documentation

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
* [x] I'm merging against the `development` branch.
* [x] I ran `cargo-fmt --all` before pushing.
* [x] I ran `cargo test` successfully before submitting my PR.
* [x] I have squashed my commits into a single commit.
* [ ] My change requires a change to the documentation.
* [ ] I have updated the documentation accordingly.
* [x] I have added tests to cover my changes.

philipr-za reviewed (dismissed):
Not hugely familiar with the context of this PR but LGTM

stringhandler commented on base_layer/core/src/base_node/service/blockchain_state/service.rs:
```diff
@@ -0,0 +1,211 @@
+            )),
+            FindFirstMatchingHeader(hashes, reply) => Box::pin(reply_or_cancel(
+                reply,
+                Self::find_first_matching_header(self.db(), hashes),
```
In my opinion this should start sending blocks or headers back immediately. The rational is that a malicious node could advertise a false tip height, then on request, it can return a valid header and then ghost you on the next call.

If the sender must immediately provide a chain, it must have created a fake chain in order to respond to this request. This reduces the attack surface significantly.

sdbondi replied:
Is that a big problem? The RPC protocol will time out and we can move onto another peer (either way we must implement this). They could respond immediately with one header, then not stream the rest (essentially the same thing), so I don't think we can prevent peers from 'ghosting'.  

I'm not sure about this though, my thinking is that we establish the starting point for the sync and that the peer is willing and able to provide the blocks. Then begin the stream. I think that there is no reason in particular that we cannot start streaming a few headers from this call (well, the RPC frontend that uses this service call).

sdbondi replied:
How about this?

```proto
message FindChainSplitRequest {
  repeated bytes block_hashes = 1;
  uint64 count = 2;
}

message FindChainSplitResponse {
  // An ordered list of headers starting from next header after the matching hash, up until `FindChainSplitRequest::count`
  repeated tari.core.BlockHeader headers = 1;
  // The index of the hash that matched from `FindChainSplitRequest::block_hashes`. This value could also be used to know how far back a split occurs.
  uint32 found_hash_index = 2;
}

```

stringhandler commented on base_layer/core/src/base_node/service/blockchain_state/service.rs:
```diff
@@ -0,0 +1,211 @@
+
+    async fn fetch_blocks(
+        db: BlockchainDatabase<B>,
+        start: Bound<u64>,
```
Apart from local querying and GRPC, I don't see the value in using heights instead of hashes. The sync node's chain could change during this step or in between calls, making this query invalid. Perhaps pass in a starting hash and a number of blocks.

sdbondi replied:
Great idea! Will change the api.

stringhandler commented on base_layer/core/src/base_node/service/blockchain_state/service.rs:
```diff
@@ -0,0 +1,211 @@
+
+    async fn fetch_headers(
+        db: BlockchainDatabase<B>,
+        start: Bound<u64>,
```
Same note here on the heights vs using hashes

stringhandler commented on comms/src/protocol/rpc/status.rs:
```diff
@@ -157,8 +171,10 @@ pub enum RpcStatusCode {
+    NotFound = 7,
     /// Unrecognised RPC status code
-    InvalidRpcStatusCode = 7,
+    InvalidRpcStatusCode = 100,
```
```suggestion
    InvalidRpcStatusCode = 500,
```

sdbondi replied:
So this variant actually isn't a specific code, it stands for "anything else other than the above". The only reason it's there at all is because ~~we have to put a number there~~(guess not - rust is happy ;)). I should add a comment for that.

That is why moving this number to > 7 doesnt actually break anything (other than older clients will say invalid status code instead of not found). So we can make this whatever we want really, it's more what makes the code clearer. I would prefer it to be within a u8 though, it doesn't matter at all, but if in the future it does, it's better to save the byte (varint encoding).

sdbondi replied:
I think it'll actually be clearer if we leave the number off completely. Rust will auto assign it, and it's clearer that the variant doesn't represent a single code.

stringhandler reviewed (commented):
Looks good. I think the use of heights instead of hashes could let to unexpected responses though. I would also prefer the FindChainSplit to immediately return headers to increase the cost of spoofing

sdbondi committed:
Base node block sync RPC service

RPC service for faster block sync. Allows streaming of full blocks,
headers and a method for chain split detection. The client-side is next
up.

- Added BlockchainStateService. This service should eventually be the only service
  which provides a query interface to the blockchain database for other
  services.
- Added BaseNodeSyncRpcService. This RPC service exposes `sync_blocks`,
  `sync_headers`, `get_header_by_height` and `find_chain_split`. These
  functions will be used in the improved block sync.
- Added `NotFound` status to RpcStatus
- Added `RpcStatus::log_internal_error()` convenience function to make
  dealing with misc errors in RPC handler functions easier.
- Added `spawn_until_shutdown` method to the service initialiser context
   that spawns a task until the shutdown is triggered. This removes the need for
   implementers to worry about how to implement shutdown in initialisers.
- Cleaned up some test helpers
- Added unit tests for the Base node sync RPC handlers
- Added unit tests for the BlockchainStateService

Ref #2113

stringhandler reviewed (approved):
Looks good. The tests are really nice and easy to read.
